File: src/robot_bringup/scripts/cmd_vel_joint_relay.py
#!/usr/bin/env python3
"""
cmd_vel_joint_relay.py

Derives /joint_states from /cmd_vel using inverse Ackermann kinematics.
Used in hardware mode (use_gazebo:=false) to keep the TF tree complete and
animate the robot model in RViz.  FOR VISUALISATION ONLY.

Signal chain context
--------------------
In Gazebo + PX4 SITL the joint-state feed is physics-derived:

  /cmd_vel → rover_speed_steering_mode → PX4 control loop
           → gz_bridge (uORB → Gz transport: servo_0, command/motor_speed)
           → Gazebo JointPositionController / JointController (P-gain)
           → Gazebo physics engine  ← actual simulated dynamics
           → gz-sim-joint-state-publisher-system  ← reads real simulated state
           → joint_state_bridge → /joint_states

On real hardware that feedback loop does not exist:

  /cmd_vel → rover_speed_steering_mode → PX4 control loop
           → PWM/UART → ESC → motor / servo → physical joints
                                              ← NO feedback to ROS

Why /cmd_vel and not PX4 ActuatorServos / ActuatorMotors:
  - Those are fmu/in/ topics (external → PX4 override), NOT PX4 output topics.
    PX4 does not publish its actuator state via DDS; the gz_bridge path is fully
    internal (uORB → Gz transport, no DDS involved).
  - Even if repurposed as feedback they are still upstream of ESC/servo response
    ("commanded, not measured") — same principle as /cmd_vel but with added DDS
    dependency and uncalibrated scaling (normalised [-1,1] → rad/rpm).
  - /cmd_vel works when only cameras + Nav2 are running (no PX4 DDS required).

Kinematics (per 50 Hz publish cycle)
-------------------------------------
  R      = linear.x / angular.z          (signed turn radius)
  δ_left  = atan2(L, R - T/2)            (inner / outer kingpin, full Ackermann)
  δ_right = atan2(L, R + T/2)
  ω_wheel = linear.x / wheel_radius      (same for all 4 — single ESC channel)
  θ_wheel += ω_wheel * dt                (integrated for mesh spin animation)

Parameters
----------
wheelbase    (float, 0.174 m)    Rear-axle to front-axle distance.
track        (float, 0.174 m)    Left-to-right wheel-centre distance.
wheel_radius (float, 0.0385 m)   Wheel radius.
robot_ns     (str,  'ackermann')  Prefix used in joint names (matches URDF ns arg).
publish_rate (float, 50.0 Hz)    Joint-state publication rate.
"""

import math

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped
from sensor_msgs.msg import JointState


class CmdVelJointRelay(Node):
    def __init__(self) -> None:
        super().__init__('cmd_vel_joint_relay')

        self.declare_parameter('wheelbase',    0.174)
        self.declare_parameter('track',        0.174)
        self.declare_parameter('wheel_radius', 0.0385)
        self.declare_parameter('robot_ns',     'ackermann')
        self.declare_parameter('publish_rate', 50.0)

        L  = self.get_parameter('wheelbase').value
        T  = self.get_parameter('track').value
        r  = self.get_parameter('wheel_radius').value
        ns = self.get_parameter('robot_ns').value
        hz = self.get_parameter('publish_rate').value

        self._L, self._T, self._r = L, T, r

        # Joint name order must match ackermann_controller.yaml and the URDF.
        self._names = [
            f'{ns}/rear_left_wheel_joint',
            f'{ns}/rear_right_wheel_joint',
            f'{ns}/front_left_wheel_joint',
            f'{ns}/front_right_wheel_joint',
            f'{ns}/front_left_wheel_steering_joint',
            f'{ns}/front_right_wheel_steering_joint',
        ]

        # Accumulated wheel rotation (rad), integrated from cmd_vel.
        # Order: [rear_left, rear_right, front_left, front_right]
        self._wheel_pos = [0.0, 0.0, 0.0, 0.0]

        self._vx = 0.0   # last commanded linear.x  (m/s)
        self._wz = 0.0   # last commanded angular.z  (rad/s)
        self._last_t = self.get_clock().now()

        self._sub = self.create_subscription(
            TwistStamped, 'cmd_vel', self._on_cmd_vel_stamped, 10
        )
        self._pub = self.create_publisher(JointState, 'joint_states', 10)
        self.create_timer(1.0 / hz, self._publish)

        self.get_logger().info(
            f'cmd_vel_joint_relay ready  '
            f'(listening on /cmd_vel as TwistStamped, '
            f'L={L} m  T={T} m  r={r} m  ns={ns}  rate={hz} Hz)'
        )

    # ------------------------------------------------------------------

    def _on_cmd_vel_stamped(self, msg: TwistStamped) -> None:
        self._vx = msg.twist.linear.x
        self._wz = msg.twist.angular.z

    def _ackermann_angles(self, v: float, w: float):
        """
        Full per-kingpin Ackermann steering angles from bicycle-model cmd_vel.

        Turn radius (signed, positive = left):  R = v / w

        Left  kingpin:  delta_L = atan2(L,  R - T/2)   (inner wheel on left turn)
        Right kingpin:  delta_R = atan2(L,  R + T/2)   (outer wheel on left turn)

        atan2(L, R±T/2) handles all sign combinations of R automatically.
        Output is clamped to the URDF joint limits of ±0.6 rad.
        """
        L, T = self._L, self._T
        LIMIT = 0.6  # rad — matches URDF front_*_wheel_steering_joint limits

        if abs(w) < 1e-6:
            return 0.0, 0.0

        if abs(v) < 1e-3:
            # Pure-rotation command: kinematically ill-defined for Ackermann.
            # Apply max steering in the commanded direction; wheels stay still.
            s = math.copysign(LIMIT, w)
            return s, s

        R  = v / w
        dl = math.atan2(L, R - T / 2.0)
        dr = math.atan2(L, R + T / 2.0)
        return (
            max(-LIMIT, min(LIMIT, dl)),
            max(-LIMIT, min(LIMIT, dr)),
        )

    def _publish(self) -> None:
        now = self.get_clock().now()
        dt  = (now - self._last_t).nanoseconds * 1e-9
        self._last_t = now

        v, w = self._vx, self._wz
        r    = self._r

        # Both rear wheels commanded at the same speed (single ESC channel).
        omega_rear = v / r

        # Front wheels free-roll; approximate at rear speed for display.
        # True path-radius correction is < 2 % at max steer for this geometry.
        omega_front = v / r

        # Integrate wheel rotation positions.
        self._wheel_pos[0] += omega_rear  * dt
        self._wheel_pos[1] += omega_rear  * dt
        self._wheel_pos[2] += omega_front * dt
        self._wheel_pos[3] += omega_front * dt

        delta_l, delta_r = self._ackermann_angles(v, w)

        msg = JointState()
        msg.header.stamp = now.to_msg()
        msg.name     = self._names
        msg.position = [
            self._wheel_pos[0], self._wheel_pos[1],   # rear  l/r spin angle
            self._wheel_pos[2], self._wheel_pos[3],   # front l/r spin angle
            delta_l,            delta_r,               # steering angles
        ]
        msg.velocity = [
            omega_rear,  omega_rear,
            omega_front, omega_front,
            0.0,         0.0,
        ]
        self._pub.publish(msg)


def main() -> None:
    rclpy.init()
    node = CmdVelJointRelay()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/robot_bringup/src/cmd_vel_joint_relay.cpp
// cmd_vel_joint_relay.cpp
//
// Derives /joint_states from /cmd_vel using inverse Ackermann kinematics.
// Used in hardware mode (use_gazebo:=false) to keep the TF tree complete and
// animate the robot model in RViz. FOR VISUALISATION ONLY.
//
// Ported from cmd_vel_joint_relay.py to C++ to avoid the rclpy executor
// overhead at the timer + subscriber callback rate. Public behaviour is
// identical: same joint names, same kinematics, same default parameters
// (except publish_rate which is 10 Hz here vs 50 Hz in the Python version —
// 10 Hz is plenty for RViz wheel-mesh animation and cuts CPU ~5x).
//
// Kinematics (per publish cycle):
//   R       = linear.x / angular.z          (signed turn radius)
//   delta_L = atan2(L,  R - T/2)            (left  / inner kingpin)
//   delta_R = atan2(L,  R + T/2)            (right / outer kingpin)
//   omega   = linear.x / wheel_radius       (single ESC channel; same all 4)
//   theta_i += omega * dt                   (integrated for mesh spin)
//
// Joint name order matches ackermann_controller.yaml + URDF.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <string>

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <sensor_msgs/msg/joint_state.hpp>

namespace
{
constexpr double kSteeringLimit = 0.6;  // rad — matches URDF front_*_wheel_steering_joint limits
}

class CmdVelJointRelay : public rclcpp::Node
{
public:
  CmdVelJointRelay() : Node("cmd_vel_joint_relay")
  {
    L_ = declare_parameter<double>("wheelbase", 0.174);
    T_ = declare_parameter<double>("track", 0.174);
    r_ = declare_parameter<double>("wheel_radius", 0.0385);
    const auto ns = declare_parameter<std::string>("robot_ns", "ackermann");
    const auto hz = declare_parameter<double>("publish_rate", 10.0);

    names_ = {
      ns + "/rear_left_wheel_joint",
      ns + "/rear_right_wheel_joint",
      ns + "/front_left_wheel_joint",
      ns + "/front_right_wheel_joint",
      ns + "/front_left_wheel_steering_joint",
      ns + "/front_right_wheel_steering_joint",
    };

    last_t_ = now();

    sub_ = create_subscription<geometry_msgs::msg::TwistStamped>(
      "cmd_vel", 10,
      [this](geometry_msgs::msg::TwistStamped::ConstSharedPtr msg) {
        vx_ = msg->twist.linear.x;
        wz_ = msg->twist.angular.z;
      });

    pub_ = create_publisher<sensor_msgs::msg::JointState>("joint_states", 10);

    const auto period = std::chrono::duration<double>(1.0 / hz);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(period),
      [this]() { publish(); });

    RCLCPP_INFO(get_logger(),
      "cmd_vel_joint_relay (C++) ready — listening on /cmd_vel (TwistStamped); "
      "L=%.3f m, T=%.3f m, r=%.4f m, ns=%s, rate=%.1f Hz",
      L_, T_, r_, ns.c_str(), hz);
  }

private:
  void ackermann_angles(double v, double w, double & dl, double & dr) const
  {
    if (std::abs(w) < 1e-6) {
      dl = dr = 0.0;
      return;
    }
    if (std::abs(v) < 1e-3) {
      // Pure-rotation command: kinematically ill-defined for Ackermann.
      // Apply max steering in the commanded direction; wheels stay still.
      const double s = std::copysign(kSteeringLimit, w);
      dl = dr = s;
      return;
    }
    const double R = v / w;
    dl = std::atan2(L_, R - T_ / 2.0);
    dr = std::atan2(L_, R + T_ / 2.0);
    dl = std::clamp(dl, -kSteeringLimit, kSteeringLimit);
    dr = std::clamp(dr, -kSteeringLimit, kSteeringLimit);
  }

  void publish()
  {
    const auto t = now();
    const double dt = (t - last_t_).seconds();
    last_t_ = t;

    const double v = vx_;
    const double w = wz_;

    // Both rear wheels commanded at the same speed (single ESC channel).
    // Front wheels free-roll; approximated at rear speed for display.
    const double omega = v / r_;

    wheel_pos_[0] += omega * dt;  // rear  left
    wheel_pos_[1] += omega * dt;  // rear  right
    wheel_pos_[2] += omega * dt;  // front left
    wheel_pos_[3] += omega * dt;  // front right

    double dl = 0.0;
    double dr = 0.0;
    ackermann_angles(v, w, dl, dr);

    sensor_msgs::msg::JointState msg;
    msg.header.stamp = t;
    msg.name = names_;
    msg.position = {
      wheel_pos_[0], wheel_pos_[1], wheel_pos_[2], wheel_pos_[3],
      dl, dr,
    };
    msg.velocity = {omega, omega, omega, omega, 0.0, 0.0};
    pub_->publish(msg);
  }

  double L_ = 0.174;
  double T_ = 0.174;
  double r_ = 0.0385;

  std::vector<std::string> names_;
  std::array<double, 4> wheel_pos_{0.0, 0.0, 0.0, 0.0};

  double vx_ = 0.0;
  double wz_ = 0.0;
  rclcpp::Time last_t_;

  rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr sub_;
  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CmdVelJointRelay>());
  rclcpp::shutdown();
  return 0;
}
